src/fileformats/ocd_file_export.cpp

```diff
@@ -48,4 +884,2019 @@ void OcdFileExport::doExport()
+		if (uses_registration_color)
+			++begin_of_spot_colors;  // in ocd output (ocd_number below)
+		if (begin_of_spot_colors > 256)
+			throw FileFormatException(tr("The map contains more than 256 colors which is not supported by ocd version 8."));
+		
+		using std::begin; using std::end;
+		auto separation_info = symbol_header.separation_info;
+		auto color_info = symbol_header.color_info;
+		quint16 ocd_number = 0;
+		
+		if (uses_registration_color)
+		{
+			color_info->number = 0;
+			color_info->name = toOcdString(Map::getRegistrationColor()->getName());
+			color_info->cmyk = { 200, 200, 200, 200 };  // OC*D stores CMYK values as integers from 0-200.
+			std::fill(begin(color_info->separations), begin(color_info->separations) + symbol_header.num_separations, 200);
+			std::fill(begin(color_info->separations) + symbol_header.num_separations, end(color_info->separations), 255);
+			++color_info;
+			++ocd_number;
+		}
+		
+		for (int i = 0; i < num_colors; ++i)
+		{
+			const auto color = map->getColor(i);
+			const auto& cmyk = color->getCmyk();
+			// OC*D stores CMYK values as integers from 0-200.
+			auto ocd_cmyk = Ocd::CmykV8 {
+			                  quint8(qRound(200 * cmyk.c)),
+			                  quint8(qRound(200 * cmyk.m)),
+			                  quint8(qRound(200 * cmyk.y)),
+			                  quint8(qRound(200 * cmyk.k))
+			};
+			
+			if (color->getSpotColorMethod() == MapColor::SpotColor)
+			{
+				separation_info->name = toOcdString(color->getSpotColorName());
+				separation_info->cmyk = ocd_cmyk;
+				separation_info->raster_freq = 1500;  /// \todo Spot color raster frequency
+				separation_info->raster_angle = 450;  /// \todo Spot color raster angle
+				++separation_info;
+				
+				if (ocd_number >= begin_of_spot_colors)
+					continue;  // Just a spot color, not a regular map color.
+				
+				std::fill(begin(color_info->separations), end(color_info->separations), 255);
+				auto index = spotColorNumber(color);
+				if (index >= symbol_header.num_separations)
+					throw FileFormatException(tr("Invalid spot color."));
+				color_info->separations[index] = 200;
+			}
+			else
+			{
+				std::fill(begin(color_info->separations), end(color_info->separations), 255);
+				if (color->getSpotColorMethod() == MapColor::CustomColor)
+				{
+					for (const auto& component : color->getComponents())
+					{
+						auto index = spotColorNumber(component.spot_color);
+						if (index >= symbol_header.num_separations)
+							throw FileFormatException(tr("Invalid spot color."));
+						color_info->separations[index] = quint8(qRound(component.factor * 200));
+					}
+				}
+			}
+			
+			color_info->number = ocd_number;
+			color_info->name = toOcdString(color->getName());
+			color_info->cmyk = ocd_cmyk;
+			++color_info;
+			++ocd_number;
+		}
+		symbol_header.num_colors = ocd_number;
+	}
+}
+
+
+template<class Format>
+void OcdFileExport::exportSetup(OcdFile<Format>& /*file*/)
+{
```

Comment:
Because `exportSetup()` can do it, and `exportSymbols()`can't.

The fairly consistent pattern in the implementation of `OcdFileExport` (and import): 
From `template<class Format> void OcdFileExport::exportImplementation()` we call into template functions so that we get a template function instantiation for the actual types of the format. This is what we can do:
- If a format is significantly different, so that it needs a different algorithm:
  - We specialize the template function (as done in `template<> void OcdFileExport::exportSetup(OcdFile<Ocd::FormatV8>& file)`).
- If the same algorithm can be used for multiple formats:
  - If the differences are completely represented by the the exporter's state (by capturing version and parameter string handler), then we may just call into a non-template function (as done `template<class Format> void OcdFileExport::exportSetup(OcdFile<Format>& /*file*/)`. This template function is likely to get inlined, thus reducing/avoiding template bloat.
  - Otherwise we still can use the same algorithm but need to specialize it for the actual types of the format. This specialization is solved by function template instantiation, i.e. the generic algorithm goes directly into the template function. This is the case for `exportSymbol` where types and even the encoding of the symbol number vary over different versions.
  This doesn't mean that there couldn't be room for reducing template bloat by factoring out universal parts.

The basic course of development is to start with generic functions (to study feasability), to add test code (to maintain correctness), and to refactor from template functions into regular functions (to reduce template bloat).

Reply:
Got it. Information about the implementation plan was missing for full understanding the structure.